source/DSCResources/DSC_CMPullDistributionPoint/DSC_CMPullDistributionPoint.psm1

```diff
@@ -0,0 +1,342 @@
+            foreach ($item in $sourceDistros)
+            {
+                $convertParams = @{
+                    Hashtable = $item
+                    ClassName = 'DSC_CMPullDistributionPointSourceDP'
+                }
+
+                $convertDP += ConvertTo-AnyCimInstance @convertParams
+            }
+        }
+
+        $dpInstall = 'Present'
+    }
+    else
+    {
+        $dpInstall = 'Absent'
+    }
+
+    return @{
+        SiteCode                = $SiteCode
+        SiteServerName          = $SiteServerName
+        EnablePullDP            = $pullDP
+        SourceDistributionPoint = $convertDP
+        DPStatus                = $dpInstall
+    }
+}
+
+<#
+    .SYNOPSIS
+        This will set the desired state.
+
+    .PARAMETER SiteCode
+        Specifies the site code for Configuration Manager site.
+
+    .PARAMETER SiteServerName
+        Specifies the SiteServer to install the role on.
+
+    .PARAMETER EnablePullDP
+        Specifies if EnablePullDP is to be set to enabled or disabled.
+
+    .PARAMETER SourceDistributionPoint
+        Specifies the desired source distribution points and the DP ranking.
+        If enabling PullDP you must specify at least one Source DP.
+#>
+function Set-TargetResource
+{
+    [CmdletBinding()]
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [String]
+        $SiteCode,
+
+        [Parameter(Mandatory = $true)]
+        [String]
+        $SiteServerName,
+
+        [Parameter()]
+        [Boolean]
+        $EnablePullDP,
+
+        [Parameter()]
+        [Microsoft.Management.Infrastructure.CimInstance[]]
+        $SourceDistributionPoint
+    )
+
+    Import-ConfigMgrPowerShellModule -SiteCode $SiteCode
+    Set-Location -Path "$($SiteCode):\"
+    $state = Get-TargetResource -SiteCode $SiteCode -SiteServerName $SiteServerName
+
+    try
+    {
+        if ($state.DPStatus -eq 'Absent')
+        {
+            throw ($script:localizedData.DistroPointInstall -f $SiteServerName)
+        }
+
+        if (((($PSBoundParameters.EnablePullDP -eq $false) -or ([string]::IsNullOrEmpty($PSBoundParameters.EnablePullDP) -and
+           $state.EnablePullDP -eq $false)) -and ($PSBoundParameters.SourceDistributionPoint)) -or
+           ((($PSBoundParameters.EnablePullDP -eq $true) -and ($state.EnablePullDP -eq $false)) -and
+           ([string]::IsNullOrEmpty($PSBoundParameters.SourceDistributionPoint))))
+        {
+            throw $script:localizedData.InvalidConfig
+        }
+
+        if (($EnablePullDP -eq $true) -and ([string]::IsNullOrEmpty($PSBoundParameters.SourceDistributionPoint)))
+        {
+            throw $script:localizedData.PullDPEnabledThrow
+        }
+
+        if (($PSBoundParameters.ContainsKey('EnablePullDP')) -and ($EnablePullDP -ne $state.EnablePullDP))
+        {
+            Write-Verbose -Message $script:localizedData.EnablePullDP
+            $buildingParams += @{
+                EnablePullDP = $EnablePullDP
+            }
+        }
+
```

Comment:
Consider adding a throw if any source DP matches the siteServerName to avoid self-referencing.